sphinx_gallery/gen_rst.py

```diff
@@ -604,58 +600,113 @@ def generate_file_rst(fname, target_dir, src_dir, gallery_conf):
+    clean_modules()
+
+    example_rst = rst_notebook_cell(script_blocks, output_blocks,
+                                    file_conf, gallery_conf)
+    save_rst_notebook(example_rst, example_file.replace('.py', '.rst'),
+                      time_elapsed, gallery_conf)
+
+    save_thumbnail(image_path_template, src_file, file_conf, gallery_conf)
+
+    example_nb = jupyter_notebook(script_blocks)
+    save_notebook(example_nb, example_file.replace('.py', '.ipynb'))
+
+    if block_vars['execute_script']:
+        logger.debug("%s ran in : %.2g seconds\n", src_file, time_elapsed)
+        # Writes md5 checksum if example has build correctly
+        # not failed and was initially meant to run(no-plot shall not cache
+        # md5sum)
+        with open(example_file + '.md5', 'w') as file_checksum:
+            file_checksum.write(get_md5sum(example_file))
+
+    return intro, time_elapsed
+
+
+def rst_notebook_cell(script_blocks, output_blocks, file_conf, gallery_conf):
+    """Generates the rst string containing the script prose, code and output
+
+    Parameters
+    ----------
+    executed_blocks : list
 
-            time_elapsed += rtime
+    gallery_conf : dict
+        Sphinx-Gallery configuration dictionary
+
+    Returns
+    -------
+    out : str
+        RST notebook
+    """
+
+    # A simple example has two blocks: one for the
+    # example introduction/explanation and one for the code
+    is_example_notebook_like = len(script_blocks) > 2
+    example_rst = u""  # there can be unicode content
+    for (blabel, bcontent, lineno), code_output in zip(script_blocks, output_blocks):
+        if blabel == 'code':
 
             if not file_conf.get('line_numbers',
                                  gallery_conf.get('line_numbers', False)):
                 lineno = None
 
+            code_rst = codestr2rst(bcontent, lineno=lineno) + '\n'
             if is_example_notebook_like:
-                example_rst += codestr2rst(bcontent, lineno=lineno) + '\n'
+                example_rst += code_rst
                 example_rst += code_output
             else:
                 example_rst += code_output
                 if 'sphx-glr-script-out' in code_output:
                     # Add some vertical space after output
                     example_rst += "\n\n|\n\n"
-                example_rst += codestr2rst(bcontent, lineno=lineno) + '\n'
-
+                example_rst += code_rst
         else:
-            example_rst += bcontent + '\n\n'
+            example_rst += bcontent + '\n'
+    return example_rst
 
-    sys.argv = argv_orig
-    clean_modules()
 
-    # Writes md5 checksum if example has build correctly
-    # not failed and was initially meant to run(no-plot shall not cache md5sum)
-    if block_vars['execute_script']:
-        with open(example_file + '.md5', 'w') as file_checksum:
-            file_checksum.write(get_md5sum(example_file))
+def save_rst_notebook(example_rst, write_file, time_elapsed, gallery_conf):
```

Comment:
Not sure notebook applies here, see my other comment. I'd rather use "example" or something better.